util/logger.py

```diff
@@ -0,0 +1,133 @@
+import sys
+import multiprocessing
+import logging
+
+_default_log_level = logging.WARNING
+# _default_log_level = logging.DEBUG
+
+class SCLogger:
+    """Holds the logging instance that we use throught SeeChange.
+
+    Normal use: get the logger object with SCLogger.get(), which is a
+    stander logging logger object.  Or, just call SCLogger.debug,
+    SCLogger.info, SCLogger.warning, or SCLogger.error, which will
+    instantiate the singleton as necessary.
+
+    The single instance is held inside a SCLogger singleton.  In
+    principle, you could also make other SCLogger objects, but that's
+    not the standard way to use it.
+
+    """
+
+    _instance = None
+    _ordinal = 0
+
+    @classmethod
+    def instance( cls, midformat=None, datefmt='%Y-%m-%d %H:%M:%S', level=_default_log_level ):
+        """Return the singleton instance of SCLogger."""
+        if cls._instance is None:
+            cls._instance = cls( midformat=midformat, datefmt=datefmt, level=level )
+        return cls._instance
+
+    @classmethod
+    def get( cls ):
+        """Return the logging.Logger object."""
+        return cls.instance()._logger
+
+    @classmethod
+    def replace( cls, midformat=None, datefmt=None, level=None ):
+        """Replace the logging.Logger object with a new one.
+
+        Subsequent calls to SCLogger.get(), .info(), etc. will now
+        return the new one.  Will inherit the midformat, datefmt, and
+        level from the current logger if they aren't specified here.
+
+        See __init__ for parameters.
+
+        Returns the logging.Logger object you'd get from get().
+
+        """
+        if cls._instance is not None:
+            midformat = cls._instance.midformat if midformat is None else midformat
+            datefmt = cls._instance.datefmt if datefmt is None else datefmt
+            level = cls._instance._logger.level if level is None else level
+        else:
+            datefmt = '%Y-%m-%d %H:%M:%S' if datefmt is None else datefmt
+            level = _default_log_level if level is None else level
+        cls._instance = cls( midformat=midformat, datefmt=datefmt, level=level )
+        return cls._instance
+
+    @classmethod
+    def set_level( cls, level=_default_log_level ):
+        """Set the log level of the logging.Logger object."""
+        cls.instance()._logger.setLevel( level )
+
+    @classmethod
+    def setLevel( cls, level=_default_log_level ):
+        """Set the log level of the logging.Logger object."""
+        cls.instance()._logger.setLevel( level )
```

Comment:
Why do we need both these methods?

Reply:
This was a bit of laziness on my part.  The SCLogger class evolved a bit as I was working on it.  Originally, I had set_level, but then later when I added the info(), debug(), etc. methods, I could do a bulk replace of _logger with SCLogger.  That's when I remember that logging.logger usese setLevel and not set_level, so I just added the method here so that either would work when used in code.